10/umbraco-deploy/deployment-workflow/import-export.md

```diff
@@ -0,0 +1,440 @@
+                RawPropertyValues = nestedContentItem.RawPropertyValues
+            });
+        }
+
+        var blockValue = new BlockValue()
+        {
+            Layout = new Dictionary<string, JToken>()
+            {
+                { Constants.PropertyEditors.Aliases.BlockList, JToken.FromObject(layoutItems) }
+            },
+            ContentData = contentData
+        };
+
+        return JsonConvert.SerializeObject(blockValue, Formatting.None);
+    }
+}
+```
+
+Moving forward, other migrators may be built by HQ or the community for property editors found in community packages. We'll make them available for [use](https://www.nuget.org/packages/Umbraco.Deploy.Contrib) and [review](https://github.com/umbraco/Umbraco.Deploy.Contrib) via the `Umbraco.Deploy.Contrib` package.
+
+### Migrating from Umbraco 7
+
+The import and export feature is available from Deploy 4 (which supports Umbraco 8). It's not been ported back to Umbraco 7, hence you can't trigger an export from there in the same way.
+
+We are still able to use this feature though to help migration from Umbraco 7 to a more recent major version.
+
+We can generate an export zip file in the same format as that used by the content import/export feature. With that we can import it into Umbraco 8 or above.  And apply similar migrations to update obsolete data types and property data into newer equivalents.
+
+This is possible via code - by temporarily applying a composer to an Umbraco 7 solution to generate the export file on start-up.
+
+An example follows:
+
+[TBC - code sample]
+
+## Service details
+
+Underlying the functionality of import/export with Deploy is the import/export service, defined by the `IArtifactImportExportService`.
+
+You may have need to make use of this service directly if building something custom with the feature. For example you might want to import from or export to some custom storage.
+
+The service interface defines two methods:
+
+- `ExportArtifactsAsync` - takes a collection of artifacts and a storage provider defined by the `IArtifactExportProvider` interface. The artifacts are serialized and exported to storage.
+    - `IArtifactExportProvider` defines methods for creating streams for writing serialized artifacts or files handled by Deploy (media, templates, stylesheets etc.).
+- `ImportArtifactsAsync` - takes storage provider containing an import defined by the `IArtifactImportProvider` interface. The artifacts from storage are imported into Umbraco.
+    - `IArtifactImportProvider` defines methods for creating streams for reading serialized artifacts or files handled by Deploy (media, templates, stylesheets etc.).
+
+Implementations for `IArtifactExportProvider` and `IArtifactImportProvider` are provided for:
+- A physical directory - via `PhysicalDirectoryArtifactImportExportProvider`.
+- An Umbraco file system (`IFileSystem` - via `FileSystemArtifactImportExportProvider`.
+- A zip file - via `ZipArchiveArtifactImportExportProvider`.
```

Comment:
These provider implementations are all internal and only exposed using extension methods in the `Umbraco.Deploy.Infrastructure.Extensions` namespace, so maybe we should mention that instead?

Reply:
We do have a public `ArtifactImportExportProviderBase` implementation you can use to create your own import/export provider (this base implementation uses both interfaces, because there's a lot of shared code between importing and exporting and it kind of makes sense to always support both ways - although you're not strictly required to)...
```c#
public class MyCustomArtifactImportExportProvider : ArtifactImportExportProviderBase
{
    public override IEnumerable<string> GetArtifactSignatureFiles() => throw new NotImplementedException();
    protected override Stream CreateReadStream(string filePath) => throw new NotImplementedException();
    protected override Stream CreateWriteStream(string filePath) => throw new NotImplementedException();
}
```